mypy/semanal.py

```diff
@@ -1282,6 +1285,71 @@ def check_function_signature(self, fdef: FuncItem) -> None:
         elif len(sig.arg_types) > len(fdef.arguments):
             self.fail("Type signature has too many arguments", fdef, blocker=True)
 
+    def check_paramspec_definition(self, defn: FuncDef) -> None:
+        func = defn.type
+        assert isinstance(func, CallableType)
+
+        param_spec_var = next(
+            (var for var in func.variables if isinstance(var, ParamSpecType)), None
+        )
+        if param_spec_var is None:
+            return
+
+        args = func.var_arg()
+        kwargs = func.kw_arg()
+        if args is None and kwargs is None:
+            return  # Looks like this function does not have starred args
+
+        has_paramspec_callable = any(
+            arg.from_concatenate
+            or (arg.arg_types and isinstance(get_proper_type(arg.arg_types[0]), ParamSpecType))
+            for arg in get_proper_types(func.arg_types)
+            if isinstance(arg, CallableType)
+        )
+        if not has_paramspec_callable:
+            return  # Callable[ParamSpec, ...] was not found
+
+        args_type = args.typ if args is not None else None
+        kwargs_type = kwargs.typ if kwargs is not None else None
+
+        args_defn = next(
+            (
+                arg_def
+                for arg_def, arg_kind in zip(defn.arguments, defn.arg_kinds)
+                if arg_kind == ARG_STAR
+            ),
+            None,
+        )
+        kwargs_defn = next(
+            (
+                arg_def
+                for arg_def, arg_kind in zip(defn.arguments, defn.arg_kinds)
+                if arg_kind == ARG_STAR2
+            ),
+            None,
+        )
+
+        args_defn_type = args_defn.type_annotation if args_defn is not None else None
+        kwargs_defn_type = kwargs_defn.type_annotation if kwargs_defn is not None else None
+
+        # This may happen on invalid `ParamSpec` args / kwargs definition:
+        if not (
+            isinstance(args_defn_type, UnboundType)
+            and args_defn_type.name.endswith(".args")
+            or isinstance(kwargs_defn_type, UnboundType)
+            and kwargs_defn_type.name.endswith(".kwargs")
+        ):
+            # Looks like both `*args` and `**kwargs` are not `ParamSpec`
+            # It might be something else, skipping.
+            return
+
+        if not isinstance(args_type, ParamSpecType) or not isinstance(kwargs_type, ParamSpecType):
+            self.fail(
+                f'ParamSpec must have "*args" typed as "{param_spec_var.name}.args" and "**kwargs" typed as "{param_spec_var.name}.kwargs"',
```

Comment:
What if these refer to different ParamSpec variables? E.g. `*args: P1.args, **kwargs: P2.kwargs`? Could we catch this error as well (or perhaps we are catching it)?